docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,71 @@
+
+All async methods effectively have two entry points, or signatures. The first signature is the one present in the above code: a modopt before the return type. The second signature is a "Task-equivalent signature", described in further detail in [I.8.6.1.5 Method signatures].
+
+Async methods have a special calling convention and may not be called directly outside of other async methods. To call an async method from a sync method, callers must use the second "Task-equivalent signature".
+
+Callers may retrieve a Task-equivalent return type from an async method via calling the "Task-equivalent signature". This functionality is available in both sync and async methods.
+
+### I.8.6.1.5 Method signatures
+
+The list of relevant components is augmented to include sync vs. async method types. Async methods have some additions to normal signature compatibility.
+
+Async MethodDef entries implicitly create two member definitions: one explicit, primary definiton, and a second implicit, runtime-generated definition.
+
+The primary, mandatory definition must be present in metadata as a MethodDef. This signature must have a required custom modifier as the last custom modifier before the return type. The custom modifier must fit the following requirements:
+* If the return type is void, the custom modifier must be either `System.Threading.Task` or `System.Threading.ValueTask`.
+* If the return type is not `void`, the modifier must be to either `System.Threading.Task<T>` or `System.Threading.ValueTask<T>`. The return type must be a valid substitution for the type parameter of the custom modifier type.
+
+_[Note: async methods have the same return type conventions as sync methods. If the async method produces a System.Int32, the return type must be System.Int32.]_
+
+The second async signature is implicit and runtime-generated and is hereafter referred to as the "Task-equivalent" signature. It is generated based on the primary signature. The transformation is as follows:
+
+If the async return type is void, the return type of the Task-equivalent signature is the type of the async custom modifier.
+
+Otherwise, the Task-equivalent return type is the custom modifier type (either Task`1 or ValueTask`1), substituted with the async return type.
+
+### I.8.10.2 Method inheritance
+
+For the purposes of inheritance and hiding, both async signatures ([### I.8.6.1.5 Method signatures]) are used for hiding and overriding, but cannot be configured separately.
+
+### II.10.3.2 The .override directive
+
+Async methods participate in overrides through both definitions (both signatures). An async method with a .override overrides the target method signature, as well as the secondary "Task-equivalent" signature if applicable. An async method may also override only the "Task-equivalent" signature, if an async signature is not present on the base class. A sync method may not override an async method, even if it matches the async method's "Task-equivalent" definition.
```

Comment:
> An async method may also override only the "Task-equivalent" signature, **if an async signature is not present on the base class.**

Emphasis mine; what do you mean by async signature here? I'd like some clarity on a couple of scenarios:

1. Library B depends on A. A has a virtual `Task`-returning method `M` that `B` overrides. `A` upgrades to a new version and, when `M` is recompiled, it implicitly gets upgraded, and now has the new `async` signature. Does `B`'s override of `M` continue to work without recompile?
2. Same as 1, except this time `B` is the one that is upgraded, not `A`. What signature does the compiler provide as a `.override` for `A.M`? The `Task`-equivalent one, or the new-style one?

Reply:
I think async signature means the `async2` one. I think what it's saying is that you can't override an `async2` method with a task-based (current style) one.

This seems potentially problematic for compatibility of `virtual async` methods. Is it not possible for the runtime to generate `async` -> `async2` and/or vice versa helpers in this case?

Reply:
This is the ECMA-335 spec, so I want to remind everyone that `async` in C# does not exist. Async in this document always refers to "runtime async". All methods, including existing compiler-async, are "sync" methods from the runtime perspective.

> Library B depends on A. A has a virtual Task-returning method M that B overrides. A upgrades to a new version and, when M is recompiled, it implicitly gets upgraded, and now has the new async signature. Does B's override of M continue to work without recompile?

This is a great question. As written: this is a breaking change. Because a sync method can't override an async method, you cannot compatibly change a sync method to an async method.

I think that's wrong and probably an oversight on my part. We have to be able to do that for compatibility. I will need to write up rules on exactly what that entails.